wai-language-server/src/main.rs

```diff
@@ -49,16 +110,296 @@ impl LanguageServer for Backend {
+                        .unwrap();
+                }
+
+                Ok(None)
+            }
             _ => Err(Error::invalid_request()),
         }
     }
+    async fn did_open(&self, params: DidOpenTextDocumentParams) {
+        let uri = &params.text_document.uri;
+        let text = &params.text_document.text;
+        let version = &params.text_document.version;
+        let lang_id = &params.text_document.language_id;
+
+        self.import_map.insert(
+            uri.to_string(),
+            WAIFile {
+                file_name: uri.to_string(),
+                file_content: text.to_string(),
+            },
+        );
+
+        // self.client
+        //     .log_message(
+        //         MessageType::INFO,
+        //         format!("{:?} {:?} {:?} {:?}", uri, text, version, lang_id),
+        //     )
+        //     .await;
+
+        self.client
+            .log_message(MessageType::INFO, "file opened!")
+            .await;
+        // self.on_change(TextDocumentItem {
+        //     uri: params.text_document.uri,
+        //     text: params.text_document.text,
+        //     version: params.text_document.version,
+        // })
+        // .await
+    }
+
+    async fn did_change(&self, mut params: DidChangeTextDocumentParams) {
+        let uri = params.text_document.uri;
+        let text = std::mem::take(&mut params.content_changes[0].text);
+
+        self.client
+            .log_message(MessageType::INFO, format!("{:?}", text))
+            .await;
+
+        // self.on_change(TextDocumentItem {
+        //     uri: params.text_document.uri,
+        //     text: std::mem::take(&mut params.content_changes[0].text),
+        //     version: params.text_document.version,
+        // })
+        // .await
+    }
+
+    async fn did_save(&self, _: DidSaveTextDocumentParams) {
+        self.client
+            .log_message(MessageType::INFO, "file saved!")
+            .await;
+    }
+
+    async fn did_close(&self, _: DidCloseTextDocumentParams) {
+        self.client
+            .log_message(MessageType::INFO, "file closed!")
+            .await;
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+enum BindingLanguage {
+    Rust,
+    JavaScript,
+    Python,
+    C,
+}
+
+impl From<BindingLanguage> for String {
+    fn from(val: BindingLanguage) -> Self {
+        match val {
+            BindingLanguage::Rust => "Rust".to_string(),
+            BindingLanguage::JavaScript => "Javascript".to_string(),
+            BindingLanguage::Python => "Python".to_string(),
+            BindingLanguage::C => "C".to_string(),
+        }
+    }
+}
+
+impl From<&str> for BindingLanguage {
+    fn from(s: &str) -> Self {
+        match s {
+            "Rust" => BindingLanguage::Rust,
+            "Javascript" => BindingLanguage::JavaScript,
+            "Python" => BindingLanguage::Python,
+            "C" => BindingLanguage::C,
+            _ => panic!("Unknown binding language"),
+        }
+    }
+}
```

Comment:
You should really be implementing either the `std::str::FromStr` or `std::convert::TryFrom` traits here. That way you can handle unknown language strings gracefully instead of crashing the LSP.

Reply:
There are also custom derives that'll generate the code for you. One I've used in the past is [`strum`](https://lib.rs/strum).